apis/ui/v1alpha1/feature_types.go

```diff
@@ -0,0 +1,99 @@
+/*
+Copyright AppsCode Inc. and Contributors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	"kmodules.xyz/resource-metadata/apis/shared"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const (
+	ResourceKindFeature = "Feature"
+	ResourceFeature     = "feature"
+	ResourceFeatures    = "features"
+)
+
+// +genclient
+// +genclient:nonNamespaced
+// +k8s:openapi-gen=true
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// +kubebuilder:object:root=true
+// +kubebuilder:resource:path=features,singular=feature,scope=Cluster
+// +kubebuilder:subresource:status
+// +kubebuilder:printcolumn:name="Enabled",type="boolean",JSONPath=".status.enabled"
+// +kubebuilder:printcolumn:name="Managed",type="boolean",JSONPath=".status.managed"
+// +kubebuilder:printcolumn:name="Ready",type="boolean",JSONPath=".status.ready"
+// +kubebuilder:printcolumn:name="Age",type="date",JSONPath=".metadata.creationTimestamp"
+type Feature struct {
+	metav1.TypeMeta `json:",inline"`
+	// +optional
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+	Spec              FeatureSpec   `json:"spec,omitempty"`
+	Status            FeatureStatus `json:"status,omitempty"`
+}
+
+type FeatureSpec struct {
+	// Title specify the title of this feature.
+	Title string `json:"title"`
+	// Description specifies a short description of the service this feature provides.
+	Description string `json:"description"`
+	// Icons is an optional list of icons for an application. Icon information includes the source, size,
+	// and mime type. These icons will be used in UI.
+	Icons []shared.ImageSpec `json:"icons,omitempty"`
+	// FeatureSet specifies the name of the FeatureSet where this feature belong to.
+	FeatureSet string `json:"featureSet"`
+	// Required specify whether this feature is mandatory or not for enabling the respecting FeatureSet.
+	// +optional
+	Required bool `json:"required,omitempty"`
+	// Requirements specifies the requirements for this feature to consider enabled.
+	Requirements Requirements `json:"requirements"`
+}
+
+type Requirements struct {
+	// DependsOn specifies a list of Feature names that must be enabled for using this feature.
+	// +optional
+	DependsOn []string `json:"dependsOn,omitempty"`
+	// RequiredResources specifies the resources that should be registered to consider this feature as enabled.
+	RequiredResources []metav1.GroupVersionKind `json:"requiredResources,omitempty"`
```

Comment:
- `DependsOn` doesn't fully express its actual requirements, features. As it's a subfield of the `Required` struct, maybe this can be just `Features`.  <===> [ Requirements.Features ]
- `RequiredResources` could be just `Resources` <===> [ Requirements.Resources ]